lib/msal-common/src/client/AuthorizationCodeClient.ts

```diff
@@ -0,0 +1,165 @@
+        );
+        parameterBuilder.addScopes(scopes);
+
+        // validate the redirectUri (to be a non null value)
+        RequestValidator.validateRedirectUri(request.redirectUri);
+        parameterBuilder.addRedirectUri(request.redirectUri);
+
+        // add code: user set, not validated
+        parameterBuilder.addAuthorizationCode(request.code);
+
+        // add code_verifier if passed
+        if (request.codeVerifier) {
+            parameterBuilder.addCodeVerifier(request.codeVerifier);
+        }
+
+        parameterBuilder.addGrantType(GrantType.AUTHORIZATION_CODE_GRANT);
+
+        return parameterBuilder.createQueryString();
+    }
+
+    /**
+     * This API validates the `AuthorizationCodeUrlRequest` and creates a URL
+     * @param request
+     */
+    private createAuthCodeUrlQueryString(request: AuthorizationCodeUrlRequest): string {
+        const parameterBuilder = new RequestParameterBuilder();
+
+        parameterBuilder.addClientId(this.config.authOptions.clientId);
+
+        // validate and add scopes
+        const scopes = RequestValidator.validateAndGenerateScopes(
+            request.scopes,
+            this.config.authOptions.clientId
+        );
+        parameterBuilder.addScopes(scopes);
+
+        // validate the redirectUri (to be a non null value)
+        RequestValidator.validateRedirectUri(request.redirectUri);
+        parameterBuilder.addRedirectUri(request.redirectUri);
+
+        // generate the correlationId if not set by the user and add
+        const correlationId = request.correlationId
+            ? request.correlationId
+            : this.config.cryptoInterface.createNewGuid();
+        parameterBuilder.addCorrelationId(correlationId);
+
+        // add response_mode. If not passed in it defaults to query.
+        parameterBuilder.addResponseMode(request.responseMode);
+
+        // add response_type = code
+        parameterBuilder.addResponseTypeCode();
+
+        if (request.codeChallenge) {
+            RequestValidator.validateCodeChallengeParams(request.codeChallenge, request.codeChallengeMethod);
+            parameterBuilder.addCodeChallengeParams(request.codeChallenge, request.codeChallengeMethod);
+        }
+
+        if (request.state) {
+            parameterBuilder.addState(request.state);
+        }
+
+        if (request.prompt) {
+            RequestValidator.validatePrompt(request.prompt);
+            parameterBuilder.addPrompt(request.prompt);
+        }
+
+        if (request.loginHint) {
+            parameterBuilder.addLoginHint(request.loginHint);
+        }
+
+        if (request.domainHint) {
+            parameterBuilder.addDomainHint(request.domainHint);
+        }
+
+        if (request.nonce) {
+            parameterBuilder.addNonce(request.nonce);
+        }
+
+        if(request.claims) {
+            parameterBuilder.addClaims(request.claims);
+        }
+
+        return parameterBuilder.createQueryString();
+    }
+}
```

Comment:
These should be abstracted into a separate class. In fact, I would argue that the RequestParameterBuilder class (as it is) is an unnecessary layer of abstraction. You could pull these functions into that class and get rid of the current functions that are there.

Reply:
Not sure I follow, could you please expand? 

Are you suggesting that we move  `createTokenRequestBody` and `createQueryString` into  `RequestParameterBuilder` and abstract them so that work for every single type of request (AuthCode, DeviceCode, etc)?